File: go/findDistinctvalueinLL/findDistinctvalueinLL.go
package main

import "fmt"

type LinkedList struct {
	Val  int
	Next *LinkedList
}

// Time -> 0(N) Space -> 0(N)
func removeduplicates(head *LinkedList) *LinkedList {
	set := make(map[int]int, 0)
	var newHead *LinkedList
	actualHead := newHead
	tmpHead := head

	for head != nil {
		set[head.Val]++
		head = head.Next
	}

	head = tmpHead
	for head != nil {
		tmpVal := head.Val
		if val := set[head.Val]; val > 1 {
			for head != nil && head.Val == tmpVal {
				head = head.Next
			}
		} else {
			if newHead == nil {
				newHead = head
				actualHead = newHead
				head = head.Next
			} else {
				newHead.Next = head
				head = head.Next
				newHead = newHead.Next
			}
			newHead.Next = nil
		}
	}

	return actualHead
}

func main() {
	B := &LinkedList{Val: 2, Next: nil}
	A := &LinkedList{Val: 1, Next: B}
	head := &LinkedList{Val: 1, Next: A}
	fmt.Println(removeduplicates(head))
}

File: python/findDistinctValueinLL.py
class LinkedList:
    def __init__(self, val=0, next=None) -> None:
        self.val = val
        self.next = next


# Time -> 0(N) Space -> 0(N)
def findDistinct(head):
    set = {}
    tmpHead = head
    newHead = None
    actualHead = None

    while head is not None:
        if head.val in set.keys():
            set[head.val] += 1
        else:
            set[head.val] = 1
        head = head.next

    head = tmpHead
    while head is not None:
        tmpVal = head.val
        if set[tmpVal] > 1:
            while head is not None and head.val == tmpVal:
                head = head.next
        else:
            if newHead is None:
                newHead = head
                actualHead = newHead
                head = head.next
            else:
                newHead.next = head
                head = head.next
                newHead = newHead.next
            newHead.next = None

    return actualHead
